File: Data Structures & Algorithms/coin-change/submission-5.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int n = coins.length;
        int[] f = new int[amount + 1];
        Arrays.fill(f, Integer.MAX_VALUE / 2);
        
        // f[i+1][c] = Math.min(f[i][c] , f[i+1][c - nums[i]]+1)
        // f[c] = Math.min(f[c], f[c - nums[i]) + 1]);
        f[0] = 0;

        for (int coin : coins){
            for (int c = coin ; c <= amount; c++) {
                f[c] = Math.min(f[c], f[c - coin] + 1);
            }
        }
        int ans = f[amount];

        return ans < Integer.MAX_VALUE / 2 ? ans : -1;
    }
}

File: Data Structures & Algorithms/maximum-subarray/submission-5.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        min_pre_sum = pre_sum = 0
        ans = float('-inf')

        for n in nums:
            pre_sum += n
            ans = max(ans, pre_sum - min_pre_sum)
            min_pre_sum = min(min_pre_sum, pre_sum)
        return ans    
            